mismi-s3/src/Mismi/S3/Commands.hs

```diff
@@ -622,6 +631,26 @@ downloadWithRange a start end dest = withRetries 5 $ do
     Just () -> pure ()
     Nothing -> liftIO $ ioError (userError "downloadWithRange timeout")
 
+downloadRecursive :: Address -> FilePath -> EitherT DownloadError AWS ()
+downloadRecursive src dest = do
+  -- Check if the destination already exists and is not a directory.
+  es <- tryIO $ getFileStatus dest
+  case es of
+    Left _ -> pure ()
+    Right st -> unless (isDirectory st) . left $ DownloadDestinationNotDirectory dest
+  -- Real business starts here.
+  addrs <- lift $ listRecursively src
+  mapM_ drWorker addrs
+  where
+    tryIO :: MonadIO m => IO a -> m (Either IOError a)
+    tryIO = liftIO . CE.try
+
+    drWorker :: Address -> EitherT DownloadError AWS ()
+    drWorker addr = do
+      fpdest <- hoistMaybe (DownloadInvariant addr src) $
+                    ((</>) dest) . T.unpack . unKey <$> removeCommonPrefix src addr
+      liftIO . createDirectoryIfMissing True $ takeDirectory fpdest
+      downloadSingle addr fpdest
```

Comment:
Would it make sense to use `download` here instead of `downloadSingle` to parallelise large file downloads? (Could also add concurrency around the individual objects, but probably makes more sense to do it chunkwise).

Reply:
Yeah, @nhibberd brought this up too. This is the most simple version that works. When I have upload working I can then test upload against download and then look at use cases like lots of small files vs a small number of large files.

Reply:
Actually revising this, I should change that now.